### ISSUE TITLE
Save multiple main files

### ISSUE THREAD

**Issue by DifficultNick** [mrcrowl/vscode-easy-less#20]
In this case:
`// main: file1.less, main: file2.less`
where _file2.less_ contains `@import "file1.css"`
I think it will be nice to wait while _file1.css_ wil be saved before compile _file2.less_

**Comment by mrcrowl** [comment]
@DifficultNick I'm not sure I understand the use case in which a less file would `@import` a .css file.  This isn't something I've come across before.

Can you provide an example of a situation in which you'd like to use this?

**Comment by DifficultNick** [comment]
Oh, sorry, i was wrong. Of course _less_, but not _css_.
I meant following:
i need to **combine** a lot of small _less_ files into one big _less_ file. 
I do it this way (_combine.less_):
```
// out: big.less
@import (inline) small_1.less
@import (inline) small_2.less
```
Also I need to **compile** a file (_style.less_) like this:
```
// out: true
@import "variables";
@import "big.less"
```

So when i compile a _small.less_ with this settings:
`// main: combine.less, style.less`

_style.css_ always contains data of previous _big.less_ version.

**Comment by mrcrowl** [comment]
Okay, just to confirm, this is how I understand your situation:

![easy-less issue 20](https://cloud.githubusercontent.com/assets/1796256/25770786/df3d1e12-3292-11e7-8e99-c413079bdbdd.png)

... and you want `combine.less` to output `big.less` _before_ `style.less` imports it and outputs `style.css`?

**Comment by DifficultNick** [comment]
yes, this now works amazing!
thank you!

**Comment by DifficultNick** [comment]
It seems this problem became actual again.
Of course I can solve it by double saving (thanks to your latest release), but it would be very nice to make it in more propper way =)

**Comment by CnSimonChan** [comment]
Hi @DifficultNick,

Sorry for the inconvenience, the last update is contributed by me.

May I know one thing that, is your "big.less" opened in editor? In my test it works if I don't have "big.less" opened.

**Comment by DifficultNick** [comment]
It doesn't matter if it opened or not (both variants don't work for me).
Here is the examle:
https://drive.google.com/file/d/0B4BkdEiPDnKXSHZTSVhxZExRQWM/view?usp=sharing

By saving `small1.less` I suppose `style.css` to be the output of all compiled files, but `combine.less` imports `small1.less` before it was changed.

**Comment by mrcrowl** [comment]
Tricky.  I think this is related to the save event moving to `onWillSaveTextDocument`.  Will need to investigate further.

**Comment by CnSimonChan** [comment]
@DifficultNick Thanks for the repro, I found the problem now.

To support compiling unchanged files I had moved to `onWillSaveTextDocument` event to detect saving, but the event fires *before* the file is actually written to disk, so I need to *help* lessc find the correct imports using a plugin.

The problem is not multiple main files, but `@import (inline)`, a minimal repro should be:

a.less:
```` less
// main: b.less

a { color: red; }
````

b.less
```` less
// out: c.less

@import (inline) "a.less";
````

For normal imports lessc will pass the file to my plugin, and I can replace it with the unsaved content I read from vscode editor, but for inline imports, lessc just read the file from disk and emit it to the output.

I tried to delay my plugin in lessc plugin system a bit, then the problem is gone, but I don't understand why, technically it doesn't make any change. I will try to find another better resolution.

**Comment by mrcrowl** [comment]
@CnSimonChan thanks for the minimal repro.  Very good :)  Did you have any success with a resolution?

**Comment by CnSimonChan** [comment]
@mrcrowl In my test if I change [`pluginManager.addPreProcessor(new LessDocumentResolver());`](https://github.com/mrcrowl/vscode-easy-less/blob/1856c14a98ea3c2ace7190349b6020d033acb87b/src/LessDocumentResolverPlugin.ts#L10) to `pluginManager.addPreProcessor(new LessDocumentResolver(), 1000);`, the problem is gone.

It's very strange because the argument `1000` only changed the order of plugins executes, but they should still run before files are saved.

Can you verify it for me?

**Comment by mrcrowl** [comment]
Been looking at this further.  The problem is that a.less is never accessed by the LessDocumentResolverPlugin.ts, therefore it uses the saved version.

I think a reasonable solution is to only trigger less compile from the `willSaveEvent` when the document is clean (!isDirty).  Otherwise stick with the `didSaveEvent`.  

```typescript
    // automatically compile less on save
    const willSaveEvent = vscode.workspace.onWillSaveTextDocument(e =>
    {
        if (e.document.fileName.endsWith(LESS_EXT) && !e.document.isDirty)
        {
            new CompileLessCommand(e.document, lessDiagnosticCollection).execute()
        }
    });

    // automatically compile less on save
    const didSaveEvent = vscode.workspace.onDidSaveTextDocument(document =>
    {
        if (document.fileName.endsWith(LESS_EXT))
        {
            new CompileLessCommand(document, lessDiagnosticCollection).execute()
        }
    });
```

By keeping your LessDocumentResolverPlugin in place, this also works with unsaved files.

**Comment by mrcrowl** [comment]
@DifficultNick Fix has been released in easy-less v1.4.5
